sql-plugin/src/main/scala/com/nvidia/spark/rapids/SpillFramework.scala

```diff
@@ -0,0 +1,1606 @@
+            // to synchronize before we ask for the next chunk
+            Cuda.DEFAULT_STREAM.sync()
+          }
+        }
+      }
+      builder.build
+    }
+  }
+
+  def createHostHandleFromDeviceBuff(
+      buff: DeviceMemoryBuffer): SpillableHostBufferHandle = {
+    val handle = new SpillableHostBufferHandle(buff.getLength)
+    withResource(
+      SpillFramework.stores.hostStore.makeBuilder(handle)) { builder =>
+      builder.copyNext(buff, Cuda.DEFAULT_STREAM)
+      Cuda.DEFAULT_STREAM.sync()
+      builder.build
+    }
+  }
+}
+
+class SpillableHostBufferHandle private (
+    override val sizeInBytes: Long,
+    override var host: Option[HostMemoryBuffer] = None,
+    var disk: Option[DiskHandle] = None)
+  extends HostSpillableHandle[HostMemoryBuffer] {
+
+  override def spillable: Boolean = synchronized {
+    if (super.spillable) {
+      host.getOrElse {
+        throw new IllegalStateException(
+          s"$this is spillable but it doesn't have a materialized host buffer!")
+      }.getRefCount == 1
+    } else {
+      false
+    }
+  }
+
+  def materialize: HostMemoryBuffer = {
+    var materialized: HostMemoryBuffer = null
+    var diskHandle: DiskHandle = null
+    synchronized {
+      if (host.isDefined) {
+        materialized = host.get
+        materialized.incRefCount()
+      } else if (disk.isDefined) {
+        diskHandle = disk.get
+      } else {
+        throw new IllegalStateException(
+          "attempting to materialize a closed handle")
+      }
+    }
+    if (materialized == null) {
+      materialized = closeOnExcept(HostMemoryBuffer.allocate(sizeInBytes)) { hmb =>
+        diskHandle.materializeToHostMemoryBuffer(hmb)
+        hmb
+      }
+    }
+    materialized
+  }
+
+  override def spill: Long = {
+    if (!spillable) {
+      0L
+    } else {
+      synchronized {
+        if (disk.isEmpty) {
+          withResource(DiskHandleStore.makeBuilder) { diskHandleBuilder =>
+            val outputChannel = diskHandleBuilder.getChannel
+            GpuTaskMetrics.get.spillToDiskTime {
+              withResource(getHostBuffer.get) { hmb =>
+                val iter = new HostByteBufferIterator(hmb)
+                iter.foreach { bb =>
+                  try {
+                    while (bb.hasRemaining) {
+                      outputChannel.write(bb)
+                    }
+                  } finally {
+                    RapidsStorageUtils.dispose(bb)
+                  }
+                }
+              }
+            }
+            disk = Some(diskHandleBuilder.build)
+            disk
+          }
+        } else {
+          None
+        }
+      }
+      releaseHostResource()
+      sizeInBytes
+    }
+  }
+
+  private def releaseHostResource(): Unit = {
+    SpillFramework.removeFromHostStore(this)
+    synchronized {
+      host.foreach(_.close())
+      host = None
+    }
+  }
+  
+  def getHostBuffer: Option[HostMemoryBuffer] = synchronized {
```

Comment:
```suggestion
  private def getHostBuffer: Option[HostMemoryBuffer] = synchronized {
```
Although I'm not sure this method is worth being separate given it's only called in one place and that place blindly assumes the result is not None.

Reply:
made this `private[spill]` by mistake before, making it just private.